### PR TITLE
feat(comp:alert): add alert component

### PR DESCRIPTION
**Pull request by brenner8023**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/IDuxFE/idux/blob/main/packages/site/src/docs/Contributing.zh.md#commit)
- [x] Tests for the changes have been added/updated or not needed
- [x] Docs and demo have been added/updated or not needed

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Component style update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Application (the showcase website) / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

add Alert component

#### Props

| 名称 | 说明 | 类型  | 默认值 | 全局配置 | 备注 |
| --- | --- | --- | --- | --- | --- |
| `closable` | 信息提示是否可关闭 | `boolean` | `false` | ✅ |- |
| `closeIcon` | 自定义关闭按钮 | `string \| #closeIcon` | `close` | - | - |
| `description` | 辅助性文字介绍 | `string \| #description` | - | - |- |
| `icon` | 自定义图标 | `string \| #icon` | - | ✅ | 若要隐藏图标则传空串 |
| `showPagination` | 是否开启分页切换效果 | `boolean` | `false` | - | - |
| `type` | 设置提示类型 | `'success' \| 'info' \| 'warning' \| 'error'` | `info` | - |- |
| `title` | 信息提示内容 | `string \| string[] \| #default` | - | - |- |
| `onBeforeClose` | 关闭提示前会触发的回调函数 | `() => boolean \| Promise<boolean>` | - | - | - |
| `onClose` | 关闭提示会触发的回调函数 | `() => void` | - | - | - |

### CODE REVIEW

File: packages/components/alert/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+export const alertProps = {
+  type: IxPropTypes.oneOf<AlertType>(['success', 'info', 'warning', 'error']).def('info'),
+  closable: IxPropTypes.bool.def(false),
+  closeIcon: IxPropTypes.string.def('close'),
+  icon: IxPropTypes.string,
+  title: IxPropTypes.string,
+  description: IxPropTypes.string,
+  transitionName: IxPropTypes.string,
+  onBeforeClose: IxPropTypes.func<() => boolean | Promise<boolean>>().def(() => true),
```
Comment (danranVm) [inline packages/components/alert/src/types.ts:30]:
```suggestion
  onBeforeClose: IxPropTypes.emit<() => boolean | Promise<boolean> | void>(),
```

File: packages/components/alert/style/mixin.less [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.alert-icon-color (@type, @color) {
```
Comment (danranVm) [inline packages/components/alert/style/mixin.less:1]: 这两个 mixin 是不是可以合并成一个？

File: packages/components/alert/src/Alert.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      return
+    }
+    pageIndex.value -= 1
+  }
+  const handlePageRight = () => {
+    if (isRightDisabled.value) {
+      return
+    }
+    pageIndex.value += 1
+  }
```
Comment (danranVm) [inline packages/components/alert/src/Alert.tsx:175]:
```suggestion
  const [pageIndex, setPagIndex ] = useState(0)
  const [pageTotal, setPageTotal ] = useState(1)
  const isShowPagination = computed(() => pageTotal.value > 1)
  const isLeftDisabled = computed(() => pageIndex.value <= 0)
  const isRightDisabled = computed(() => pageIndex.value + 1 >= pageTotal.value)
  const pageText = computed(() => `${pageIndex.value + 1}/${pageTotal.value}`)

  const offsetPageIndex = (offset: -1 | 1) => {
    if (offset === -1 && isLeftDisabled.value) {
      return
    }
    if (offset === 1 && isRightDisabled.value) {
      return
    }
    pageIndex.value += offset
  }
```

**Review by danranVm — CHANGES_REQUESTED** [review]

File: packages/components/alert/src/Alert.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      paginationClass,
+      pageTextClass,
+      leftIconClass,
+      rightIconClass,
+    } = useClass(props, hasDescription, isLeftDisabled, isRightDisabled)
+    const { mergedClosable, handleCloseClick } = useClose(props, alertConfig.closable, isVisible)
+
+    return () => {
+      const children = slots.default?.() ?? []
+      setPageTotal(children.length)
```
Comment (danranVm) [inline packages/components/alert/src/Alert.tsx:51]:
```suggestion
  const chilren = computed(()=> slots.default?.() || NoopArray)
  const pageTotal = computed(() => chilren.value.length)
```
这里可以改成这样， render 函数外面用 computed 缓存起来

File: packages/components/alert/__tests__/alert.spec.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      template: `
+        <ix-alert :closable="closable" @close="onClose"></ix-alert>
+      `,
+      setup() {
+        return {
+          closable,
+          onClose,
+        }
+      },
+    })
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/__tests__/alert.spec.ts]: fix

File: packages/components/alert/__tests__/alert.spec.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    expect(wrapper.find('.ix-alert-close-icon').exists()).toBeFalsy()
+
+    closable.value = true
+    await nextTick()
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/__tests__/alert.spec.ts]: fix

File: packages/components/alert/__tests__/alert.spec.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+          :on-before-close="onBeforeClose"
+          @close="onClose"></ix-alert>
+      `,
+      setup() {
+        return {
+          onBeforeClose,
+          onClose,
+        }
+      },
+    })
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/__tests__/alert.spec.ts]: fix

File: packages/components/alert/__tests__/alert.spec.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  test('multiple message work', async () => {
+    const wrapper = mount({
+      components: { IxAlert },
+      template: `
+        <ix-alert>
+          <div>message1</div>
+          <div>message2</div>
+        </ix-alert>
+      `,
+    })
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/__tests__/alert.spec.ts]: fix

File: packages/components/alert/demo/Animation.vue [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<template>
+  <IxAlert closable transition-name="fade">我之前还没发现，当我发现的时候我已经发现了。</IxAlert>
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/demo/Animation.vue]: 已改成内置一种

File: packages/components/alert/src/Alert.tsx [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+      const title = isShowPagination.value ? children[pageIndex.value] : props.title || children
+
+      const description = slots.description?.() ?? props.description
+      hasDescription.value = !!description
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/src/Alert.tsx]: 已删除`hasDescription`

File: packages/components/alert/src/Alert.tsx [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+  const mergedClosable = computed(() => props.closable || globalClosable)
+
+  const handleCloseClick = () => {
+    Promise.resolve(props.onBeforeClose()).then(res => {
+      if (res === false) {
+        return
+      }
+      isVisible.value = false
+      callEmit(props.onClose)
+    })
```
[the remark this answers is not in the document]
Comment (brenner8023) [reply packages/components/alert/src/Alert.tsx]: 已修改

File: packages/components/alert/src/types.ts (hunk 1 above)
Comment (brenner8023) [reply packages/components/alert/src/types.ts]: fix

File: packages/components/alert/style/mixin.less (hunk 2 above)
Comment (brenner8023) [reply packages/components/alert/style/mixin.less]: 合并成为`alert-status-color`

File: packages/components/alert/src/Alert.tsx (hunk 3 above)
Comment (brenner8023) [reply packages/components/alert/src/Alert.tsx]: 这很react，已修改

File: packages/components/alert/src/Alert.tsx (hunk 4 above)
Comment (brenner8023) [reply packages/components/alert/src/Alert.tsx]: 改成了`const children = computed(() => flattenNode(slots.default?.() ?? []))`

**Review by danranVm — APPROVED** [review]

**Comment by danranVm** [comment]
需要解决一下 index.ts 的冲突，以及在 components.less 中，添加一下组件样式入口文件。

**Review by danranVm — APPROVED** [review]

**Pull request merged**